OneCommunityGlobal/HGNRest
# Roberto fix assigning duplicate badges

# Description
Whenever trying to manually assign badges to a user and duplicate badges are involved there aren't any clear updates being made. There was no logic to handle duplicate badges which would not allow badges to be updated anymore.

## Related PRS (if any):
Use latest frontend

## Main changes explained:
- Shortened fillEarnedDateToMatchCount function to handle all logic itself not needing a second helper function, also added edge case of user trying to lower badges count
- Added logic to assignBadge function to handle instances of duplicate badges and merge them to one with updated variables.


## How to test:
1. check into current branch
2. go to `src/controllers/userProfileController.js`, comment out lines `579 - 583`. *This step is important to clear backend cache, clearing cache through browser won't help when making changes to mongo compass and wanting to see changes immediately*
3. do `npm install` and `npm run build && npm start` to run this PR locally
4. log as admin user
5. go to any user profile and make sure they have at least 2 different badges assigned to them
6. create duplicate badges, here's a quick way to do this: 

https://github.com/OneCommunityGlobal/HGNRest/assets/117053202/905fe9e0-e944-4179-b893-6de73f75956e


8. Verify that either assigning a new badge or even incrementing count of any badge results in duplicate badges being merged.

## Screenshots or videos of changes:

https://github.com/OneCommunityGlobal/HGNRest/assets/117053202/1b91d657-117e-416d-a9e2-eda76d262cda



## Note:
*Make sure to follow step 2, if you make changes to the database through mongo compass, you won't see immediate changes in frontend due to the backend cache still not being renewed*
